posts/2021/02/pytorch-tensoriterator-internals-update.md

```diff
@@ -0,0 +1,321 @@
+  // dimension `dim`.
+
+  // These are the byte strides that separate each vector that spans across
+  // dimension `dim`
+  auto* result_data_bytes = data[0];
+  const auto* self_data_bytes = data[1];
+
+  for (int64_t vector_idx = 0; vector_idx < n; ++vector_idx) {
+
+    // Calculate cumulative sum for each element of the vector
+    auto cumulative_sum = (at::acc_type<float, false>) 0;
+    for (int64_t elem_idx = 0; elem_idx < self_dim_size; ++elem_idx) {
+      const auto* self_data = reinterpret_cast<const float*>(self_data_bytes);
+      auto* result_data = reinterpret_cast<float*>(result_data_bytes);
+      cumulative_sum += self_data[elem_idx * self_dim_stride];
+      result_data[elem_idx * result_dim_stride] = (float)cumulative_sum;
+    }
+
+    // Go to the next vector
+    result_data_bytes += strides[0];
+    self_data_bytes += strides[1];
+  }
+};
+
+iter.for_each(loop);
+```
+
+### Helper functions
+
+There are many helper functions within PyTorch that can simplify the creation
+and execution of a `TensorIterator`. We cannot cover all of them in this blog
+post, so you would need to discover them on your own. However, let's discuss
+one of the most common ones:
+[`make_reduction`](https://github.com/pytorch/pytorch/blob/d9e6750759b78c68e7d98b80202c67bea7ba24ec/aten/src/ATen/native/ReduceOpsUtils.h#L196)
+
+`make_reduction` creates a `TensorIterator` specifically for a reduction
+operation with one input and one output. It handles all of the
+`TensorIteratorConfig` setup internally, so we don't need to write as much
+boiler-plate code.
+
+The following example uses `make_reduction` to create a `TensorIterator` which
+is used to calculate the sum reduction of a 2-D input across dimension 1. This
+is equivalent to `torch.sum(self, dim=1)` in Python. If we didn't use
+`make_reduction`, this code would be a bit more complex and more difficult to
+write.
+
+```cpp
+at::Tensor self = at::randn({10, 10});
+int64_t dim = 1;
+bool keepdim = false;
+
+// `make_reduction` will allocate result Tensor for us, so we
+// can leave it undefined
+at::Tensor result;
+
+auto iter = at::native::make_reduction(
+  "sum_reduce",
+  result,
+  self,
+  dim,
+  keepdim,
+  self.scalar_type());
+
+// Sum reduce data from input into output
+auto sum_reduce_loop = [](char** data, const int64_t* strides, int64_t n) {
+  auto* out_data = data[0];
+  auto* in_data = data[1];
+
+  *reinterpret_cast<float*>(out_data) = 0;
+
+  for (int64_t i = 0; i < n; i++) {
+    // assume float data type for this example
+    *reinterpret_cast<float*>(out_data) += *reinterpret_cast<float*>(in_data);
+    in_data += strides[1];
+  }
+};
+
+iter.for_each(sum_reduce_loop);
+```
+
+# Conclusion
+
+This post was a very short introduction to what `TensorIterator` is actually
+capable of. If you want to learn more about how it works and what goes into
+things like collapsing the tensor size for optimizing memory access, a good
```

Comment:
I think it will be of much help to later on cover the CUDA-related TensorIterator parts.

Reply:
I agree that it would be good to eventually cover CUDA. But maybe it would be best to do that in another blog post to make things more easily digestible?

Reply:
Sure, I totally agree!